pyacq/core/manager.py

```diff
@@ -47,188 +71,127 @@ class Manager(RPCServer):
         self._next_nodegroup_name = 0
         self._next_node_name = 0
         
-        # shared socket for all RPC client connections
-        self._rpc_socket = RPCClientSocket()
-    
-    def connect_host(self, name, addr):
-        """Connect the manager to a Host.
-        
-        Hosts are used as a stable service on remote machines from which new
-        Nodegroups can be spawned or closed.
-        """
-        if name not in self.hosts:
-            hp = Manager._Host(name, addr)
-            self.hosts[name] = hp
+        # publish with the RPC server if there is one
+        server = RPCServer.get_server()
+        if server is not None:
+            server['manager'] = self
+            
+        # If the manager shuts down, then all spawned nodegroups should be
+        # closed as well.
+        atexit.register(self.close)
 
-    def disconnect_host(self, name):
-        """Disconnect the Manager from the Host identified by *name*.
+    def get_logger_info(self):
+        """Return the address of the log server and the level of the root logger.
         """
-        for ng in self.hosts[name]:
-            self.nodegroups.pop(ng.name)
-        self.hosts.pop(name)
+        return rpc_log.get_logger_address(), logging.getLogger().level
     
-    def default_host(self):
-        """Return the RPC name and address of a default Host created by the
-        Manager.
-        """
-        if self._default_host is None:
-            addr = self._addr.rpartition(b':')[0] + b':*'
-            proc = ProcessSpawner(Host, name='default-host', addr=addr)
-            self._default_host = proc
-            self.connect_host(proc.name, proc.addr)
-        return self._default_host.name, self._default_host.addr
-    
-    def close_host(self, name):
-        """Close the Host identified by *name*.
-        """
-        self.hosts[name].client.close()
-    
-    def close(self):
-        """Close the Manager.
+    def get_host(self, addr):
+        """Connect the manager to a Host's RPC server and return a proxy to the
+        host.
         
-        If a default host was created by this Manager, then it will be closed 
-        as well.
+        Hosts are used as a stable service on remote machines from which new
+        Nodegroups can be spawned or closed.
         """
-        if self._default_host is not None:
-            self._default_host.stop()
-        RPCServer.close(self)
+        if addr not in self.hosts:
+            try:
+                cli = RPCClient.get_client(addr)
+            except TimeoutError:
+                raise TimeoutError("No response from host at %s" % addr)
+                
+            try:
+                self.hosts[addr] = cli['host']
+            except KeyError:
+                raise ValueError("Contacted %s, but found no Host there." % addr)
+        return self.hosts[addr]
+    
+    def disconnect_host(self, host):
+        host.close_nodegroups(self)
```

Comment:
No--a Host can be used simultaneously by more than one Manager, so we ask it to close only the nodegroups that belong to this Manager.